rust/template/differential_datalog/src/api/c_api.rs

```diff
@@ -854,19 +869,168 @@ pub unsafe extern "C" fn ddlog_enable_timely_profiling(
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn ddlog_profile(prog: *const HDDlog) -> *const raw::c_char {
+pub unsafe extern "C" fn ddlog_dump_profile(
+    prog: *const HDDlog,
+    label: *const raw::c_char,
+) -> *const raw::c_char {
+    if prog.is_null() {
+        return ptr::null();
+    }
+    let prog = &*prog;
+    let label = if label.is_null() {
+        None
+    } else {
+        match CStr::from_ptr(label).to_str() {
+            Ok(s) => Some(s.to_string()),
+            Err(e) => {
+                prog.eprintln(&format!("invalid label string: {}", e));
+                return ptr::null_mut();
+            }
+        }
+    };
+
+    let prof_file = prog
+        .dump_profile(label.as_deref())
+        .unwrap_or_else(|e| format!("Failed to dump profile: {}", e));
+    CString::new(prof_file)
+        .map(CString::into_raw)
+        .unwrap_or_else(|e| {
+            prog.eprintln(&format!(
+                "Failed to convert profile filename to a C string: {}",
+                e
+            ));
+            ptr::null_mut()
+        })
+}
```

Comment:
This can be simplified by taking and returning `Option<NonNull<_>>`s
```suggestion
pub unsafe extern "C" fn ddlog_dump_profile(
    prog: Option<NonNull<HDDlog>>,
    label: Option<NonNull<raw::c_char>>,
) -> Option<NonNull<raw::c_char>> {
    let prog = prog?.as_ref();
    let label = CStr::from_ptr(label?)
        .map_err(|err| prog.eprintln(&format!("invalid label string: {}", err)))
        .ok()?;

    let prof_file = prog
        .dump_profile(label.as_deref())
        .map_err(|err| prog.eprintln(&format!("Failed to dump profile: {}", err)))
        .ok()?;

    match CString::new(prof_file).map(CString::into_raw) {
        Ok(string) => NonNull::new(string),
        Err(err) => {
            prog.eprintln(&format!(
                "Failed to convert profile filename to a C string: {}",
                err,
            ));
            None
        }
    }
}
```

Reply:
Nice!  We should apply this consistently throughout the API.